dotnet/src/Connectors/VectorData.Abstractions/ConnectorSupport/VectorStoreRecordModel.cs

```diff
@@ -0,0 +1,240 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Extensions.VectorData.ConnectorSupport;
+
+/// <summary>
+/// A model representing a record in a vector store collection.
+/// </summary>
+[Experimental("MEVD9001")]
+public sealed class VectorStoreRecordModel
+{
+    [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]
+    private readonly Type _recordClrType;
+
+    private VectorStoreRecordKeyPropertyModel? _singleKeyProperty;
+    private VectorStoreRecordVectorPropertyModel? _singleVectorProperty;
+    private VectorStoreRecordDataPropertyModel? _singleFullTextSearchProperty;
+
+    /// <summary>
+    /// The key properties of the record.
+    /// </summary>
+    public IReadOnlyList<VectorStoreRecordKeyPropertyModel> KeyProperties { get; }
+
+    /// <summary>
+    /// The dataproperties of the record.
+    /// </summary>
+    public IReadOnlyList<VectorStoreRecordDataPropertyModel> DataProperties { get; }
+
+    /// <summary>
+    /// The vector properties of the record.
+    /// </summary>
+    public IReadOnlyList<VectorStoreRecordVectorPropertyModel> VectorProperties { get; }
+
+    /// <summary>
+    /// All properties of the record, of all types.
+    /// </summary>
+    public IReadOnlyList<VectorStoreRecordPropertyModel> Properties { get; }
+
+    /// <summary>
+    /// All properties of the record, of all types, indexed by their model name.
+    /// </summary>
+    public IReadOnlyDictionary<string, VectorStoreRecordPropertyModel> PropertyMap { get; }
+
+    internal VectorStoreRecordModel(
+        Type recordClrType,
+        IReadOnlyList<VectorStoreRecordKeyPropertyModel> keyProperties,
+        IReadOnlyList<VectorStoreRecordDataPropertyModel> dataProperties,
+        IReadOnlyList<VectorStoreRecordVectorPropertyModel> vectorProperties,
+        IReadOnlyDictionary<string, VectorStoreRecordPropertyModel> propertyMap)
+    {
+        this._recordClrType = recordClrType;
+        this.KeyProperties = keyProperties;
+        this.DataProperties = dataProperties;
+        this.VectorProperties = vectorProperties;
+        this.PropertyMap = propertyMap;
+        this.Properties = propertyMap.Values.ToList();
```

Comment:
Does any of the current usages of the `Properties` take advantage of `IReadOnlyList`? I wonder if we should simply return `IEnumerable` instead and avoid creating the copy here.

```diff
- public IReadOnlyList<VectorStoreRecordPropertyModel> Properties { get; }
+ public IEnumerable<VectorStoreRecordPropertyModel> Properties => PropertyMap.Values;
```